# Проект "Предсказание вероятности подключения услуги"

## Постановка задачи

#### Задача

Требуется на основании имеющихся данных об абонентах Мегафон построить  алгоритм, который для каждой пары пользователь-услуга определит вероятность подключения услуги.

#### Целевая переменная

**target** - целевая переменная, где 1 означает подключение услуги, 0 - абонент не подключил услугу соответственно. 

#### Метрика качества

F1-score (sklearn.metrics.f1_score)

#### Требования к решению

*Целевая метрика*

* Метрика оценивается невзвешенным образом, , как например делает функция sklearn.metrics.f1_score(…, average=’macro’)

*Решение должно содержать*

1.	Работающая модель в формате pickle, которая принимает файл data_test.csv из корневой папки и записывает в эту же папку файл answers_test.csv. В этом файле должны находится 4 столбца: buy_time, id, vas_id и target. Target можно записать как вероятность подключения услуги.
2.	Тетрадка Jupyter Notebook с кодом модели. 
3.	Презентация в формате .pdf, в которой необходимо отразить:

   * Информация о модели, ее параметрах, особенностях и основных результатах.
   * Обоснование выбора модели и ее сравнение с альтернативами.
   * Принцип составления индивидуальных предложений для выбранных абонентов.

   Рекомендуемое количество слайдов – 5 – 10.

Файл answers_test.csv с результатами работы модели, презентацию, ноутбуки и резюме необходимо прикрепить ко второму уроку “курсовой проект”.

## Решение

### Обзор данных

В качестве исходных данных представлена информация об отклике абонентов на предложение подключения одной из услуг. Каждому пользователю может быть сделано несколько предложений в разное время, каждое из которых он может или принять, или отклонить.

Отдельным набором данных является нормализованный анонимизированный набор признаков, характеризующий профиль потребления абонента. Эти данные привязаны к определенному времени, поскольку профиль абонента может меняться с течением времени.

Данные train и test разбиты по периодам – на train доступно 4 месяцев, а на test отложен последующий месяц. 

Итого, в качестве входных данных представлены:
* **data_train.csv**: 
    * id, 
    * vas_id, 
    * buy_time, 
    * target
* **features.csv.zip**: 
    * id, 
    * ..feature_list..

И тестовый набор:
* **data_test.csv**: 
    * id, 
    * vas_id, 
    * buy_time

**Описание датасета**
* **id** - идентификатор абонента
* **vas_id** - подключаемая услуга
* **buy_time** - время покупки, представлено в формате timestamp, для работы с этим столбцом понадобится функция datetime.fromtimestamp из модуля datetime.
* **target** - целевая переменная, где 1 означает подключение услуги, 0 - абонент не подключил услугу соответственно.

#### Подключение библиотек и скриптов

In [1]:
import pandas as pd; pd.set_option('display.max_columns', None)
import dask.dataframe as dd
from dask.distributed import Client, progress

#### Загрузка данных

Загрузим исходные .csv-файлы в dask.dataframe и посмотрим состав столбцов.

In [2]:
features_dask = dd.read_csv('features.csv',sep='\t')
train_dask = dd.read_csv('data_train.csv')
test_dask = dd.read_csv('data_test.csv')

In [3]:
train_dask.head()

Unnamed: 0       id  vas_id    buy_time  target
0           0   540968     8.0  1537131600     0.0
1           1  1454121     4.0  1531688400     0.0
2           2  2458816     1.0  1534107600     0.0
3           3  3535012     5.0  1535922000     0.0
4           4  1693214     1.0  1535922000     0.0

In [4]:
test_dask.head()

Unnamed: 0       id  vas_id    buy_time
0           0  3130519     2.0  1548018000
1           1  2000860     4.0  1548018000
2           2  1099444     2.0  1546808400
3           3  1343255     5.0  1547413200
4           4  1277040     2.0  1546808400

In [5]:
features_dask.head()

Unnamed: 0       id    buy_time          0           1           2  \
0           0  2013026  1531688400  18.910029   46.980888    4.969214   
1           1  2014722  1539550800  36.690029  152.400888  448.069214   
2           2  2015199  1545598800 -67.019971  157.050888  -63.180786   
3           3  2021765  1534107600   7.010029  150.200888   -6.930786   
4           4  2027465  1533502800 -90.439971  134.220888 -104.380786   

            3           4          5         6          7         8  \
0   -1.386798    3.791754  -14.01179 -16.08618 -65.076097 -6.783660   
1  563.833202  463.841754  568.99821 -16.08618 -53.216097 -6.783660   
2  178.103202  -68.598246  156.99821   3.51382  25.183903 -0.417924   
3  216.213202   76.621754  351.84821 -16.08618 -65.076097 -6.783660   
4  153.643202 -109.798246  132.53821 -16.08618 -65.076097 -6.783660   

           9        10        11        12         13        14        15  \
0 -30.006538 -2.736081 -4.007526 -2.558912  49.520873  38.19189 -0.000725   
1 -26.544905 -2.736081 -4.007526 -2.558912  67.300873  55.97189 -0.000725   
2  -1.689642 -2.736081  9.226737 -2.558912 -66.189127 -69.87811 -0.000725   
3 -30.006538 -2.736081 -4.007526 -2.558912  37.620873  26.29189 -0.000725   
4 -30.006538 -2.736081 -4.007526 -2.558912 -66.189127 -77.51811 -0.000725   

         16        17        18        19          20         21        22  \
0 -0.016435 -0.107041  -1.17746 -3.178521  -13.940815 -10.744164 -0.094251   
1 -0.016435 -0.107041  15.77254 -3.178521  411.379185 -10.744164 -0.094251   
2 -0.016435 -0.107041  -5.41746 -0.638521    3.839185   8.855836 -0.094251   
3 -0.016435 -0.107041  83.55254 -3.178521  -13.940815 -10.744164 -0.094251   
4 -0.016435 -0.107041  -5.41746  3.181479  -13.940815 -10.744164 -0.094251   

         23        24          25        26       27        28        29  \
0 -0.001733 -0.009327   -2.082209  0.200138 -0.00909 -0.351862 -0.214366   
1 -0.001733 -0.009327  131.407791  0.200138 -0.00909 -0.351862 -0.214366   
2 -0.001733 -0.009327    0.457791  0.200138 -0.00909  0.648138  0.785634   
3 -0.001733 -0.009327   -2.082209  0.200138 -0.00909 -0.351862 -0.214366   
4 -0.001733 -0.009327   -2.082209  0.200138 -0.00909 -0.351862 -0.214366   

         30        31        32       33        34        35        36  \
0 -0.211608 -0.001884 -0.000023 -0.00003  -2.65939 -0.065583 -0.700765   
1 -0.211608 -0.001884 -0.000023 -0.00003  -2.65939 -0.065583 -0.700765   
2  0.788392 -0.001884 -0.000023 -0.00003  16.94061 -0.065583  1.839235   
3 -0.211608 -0.001884 -0.000023 -0.00003  -2.65939 -0.065583 -0.700765   
4 -0.211608 -0.001884 -0.000023 -0.00003  -2.65939 -0.065583 -0.700765   

           37            38           39           40          41  \
0  -42.026959  -2841.496068 -1085.821501 -1755.674564  -89.504287   
1 -192.026959  -2937.657200 -1181.982633 -1755.674564  447.193953   
2  507.973041  11357.393596  5126.807163  6230.586136  -89.504287   
3    7.973041  -1382.965804   363.802563 -1746.768314  -89.504287   
4  457.973041  -2942.440404 -1186.765837 -1755.674564  719.681263   

            42          43         44         45         46        47  \
0  -119.724355  -70.712019 -54.069191 -16.642826  -7.896282 -5.634035   
1  1258.981645 -119.662019 -54.602524 -65.059494  29.770382 -7.997875   
2  -119.724355  -25.128689 -24.602526  -0.526164  -7.896282 -7.917057   
3  -119.724355  -71.128686 -38.819191 -32.309494  -7.896282 -5.858933   
4  5981.273645 -119.662019 -54.602524 -65.059494  95.470388 -7.997875   

          48         49         50         51         52         53        54  \
0 -10.717958 -28.571103  57.869716  52.911014  26.828289  26.668705 -4.958702   
1 -10.717958 -28.567778  -4.130284  -8.088986  39.828289  13.668705 -3.958702   
2 -10.716588 -28.567090   4.869716   0.911014  88.828289  22.668705 -3.958702   
3 -10.717958 -28.571103 -19.130284 -21.088986 -24.171711   0.668705 -1.958702   
4 -10.717958 -28.571103   8.869716   6.91101

В датасете features.csv имеется столбец buy_time - столбец с аналогичны именем имеется в датасетах data_train.csv, data_test.csv

Загрузим исходные .csv-файлы в dask.dataframe c одновременным переименованием столбцов:
* индексов
* buy_time

In [6]:
def mapper(name):
    return {'Unnamed: 0': f'index_{name}', 
            'buy_time': f'buy_time_{name}'}


features_df = dd.read_csv('features.csv',sep='\t').rename(columns=mapper('feat'))
train_df = dd.read_csv('./data/data_train.csv').rename(columns=mapper('vas'))
test_df = dd.read_csv('./data/data_test.csv').rename(columns=mapper('vas'))

In [7]:
train_df.head()

index_vas       id  vas_id  buy_time_vas  target
0          0   540968     8.0    1537131600     0.0
1          1  1454121     4.0    1531688400     0.0
2          2  2458816     1.0    1534107600     0.0
3          3  3535012     5.0    1535922000     0.0
4          4  1693214     1.0    1535922000     0.0

In [8]:
test_df.head()

index_vas       id  vas_id  buy_time_vas
0          0  3130519     2.0    1548018000
1          1  2000860     4.0    1548018000
2          2  1099444     2.0    1546808400
3          3  1343255     5.0    1547413200
4          4  1277040     2.0    1546808400

In [9]:
features_df.head()

index_feat       id  buy_time_feat          0           1           2  \
0           0  2013026     1531688400  18.910029   46.980888    4.969214   
1           1  2014722     1539550800  36.690029  152.400888  448.069214   
2           2  2015199     1545598800 -67.019971  157.050888  -63.180786   
3           3  2021765     1534107600   7.010029  150.200888   -6.930786   
4           4  2027465     1533502800 -90.439971  134.220888 -104.380786   

            3           4          5         6          7         8  \
0   -1.386798    3.791754  -14.01179 -16.08618 -65.076097 -6.783660   
1  563.833202  463.841754  568.99821 -16.08618 -53.216097 -6.783660   
2  178.103202  -68.598246  156.99821   3.51382  25.183903 -0.417924   
3  216.213202   76.621754  351.84821 -16.08618 -65.076097 -6.783660   
4  153.643202 -109.798246  132.53821 -16.08618 -65.076097 -6.783660   

           9        10        11        12         13        14        15  \
0 -30.006538 -2.736081 -4.007526 -2.558912  49.520873  38.19189 -0.000725   
1 -26.544905 -2.736081 -4.007526 -2.558912  67.300873  55.97189 -0.000725   
2  -1.689642 -2.736081  9.226737 -2.558912 -66.189127 -69.87811 -0.000725   
3 -30.006538 -2.736081 -4.007526 -2.558912  37.620873  26.29189 -0.000725   
4 -30.006538 -2.736081 -4.007526 -2.558912 -66.189127 -77.51811 -0.000725   

         16        17        18        19          20         21        22  \
0 -0.016435 -0.107041  -1.17746 -3.178521  -13.940815 -10.744164 -0.094251   
1 -0.016435 -0.107041  15.77254 -3.178521  411.379185 -10.744164 -0.094251   
2 -0.016435 -0.107041  -5.41746 -0.638521    3.839185   8.855836 -0.094251   
3 -0.016435 -0.107041  83.55254 -3.178521  -13.940815 -10.744164 -0.094251   
4 -0.016435 -0.107041  -5.41746  3.181479  -13.940815 -10.744164 -0.094251   

         23        24          25        26       27        28        29  \
0 -0.001733 -0.009327   -2.082209  0.200138 -0.00909 -0.351862 -0.214366   
1 -0.001733 -0.009327  131.407791  0.200138 -0.00909 -0.351862 -0.214366   
2 -0.001733 -0.009327    0.457791  0.200138 -0.00909  0.648138  0.785634   
3 -0.001733 -0.009327   -2.082209  0.200138 -0.00909 -0.351862 -0.214366   
4 -0.001733 -0.009327   -2.082209  0.200138 -0.00909 -0.351862 -0.214366   

         30        31        32       33        34        35        36  \
0 -0.211608 -0.001884 -0.000023 -0.00003  -2.65939 -0.065583 -0.700765   
1 -0.211608 -0.001884 -0.000023 -0.00003  -2.65939 -0.065583 -0.700765   
2  0.788392 -0.001884 -0.000023 -0.00003  16.94061 -0.065583  1.839235   
3 -0.211608 -0.001884 -0.000023 -0.00003  -2.65939 -0.065583 -0.700765   
4 -0.211608 -0.001884 -0.000023 -0.00003  -2.65939 -0.065583 -0.700765   

           37            38           39           40          41  \
0  -42.026959  -2841.496068 -1085.821501 -1755.674564  -89.504287   
1 -192.026959  -2937.657200 -1181.982633 -1755.674564  447.193953   
2  507.973041  11357.393596  5126.807163  6230.586136  -89.504287   
3    7.973041  -1382.965804   363.802563 -1746.768314  -89.504287   
4  457.973041  -2942.440404 -1186.765837 -1755.674564  719.681263   

            42          43         44         45         46        47  \
0  -119.724355  -70.712019 -54.069191 -16.642826  -7.896282 -5.634035   
1  1258.981645 -119.662019 -54.602524 -65.059494  29.770382 -7.997875   
2  -119.724355  -25.128689 -24.602526  -0.526164  -7.896282 -7.917057   
3  -119.724355  -71.128686 -38.819191 -32.309494  -7.896282 -5.858933   
4  5981.273645 -119.662019 -54.602524 -65.059494  95.470388 -7.997875   

          48         49         50         51         52         53        54  \
0 -10.717958 -28.571103  57.869716  52.911014  26.828289  26.668705 -4.958702   
1 -10.717958 -28.567778  -4.130284  -8.088986  39.828289  13.668705 -3.958702   
2 -10.716588 -28.567090   4.869716   0.911014  88.828289  22.668705 -3.958702   
3 -10.717958 -28.571103 -19.130284 -21.088986 -24.171711   0.668705 -1.958702   
4 -10.717958 -28.571103   

In [10]:
LEN_TRAIN = len(train_df.index)
print(f'Наблюдений в "data_train.csv": {LEN_TRAIN}')

Наблюдений в "data_train.csv": 831653


In [11]:
LEN_TEST = len(test_df.index)
print(f'Наблюдений в "data_test.csv": {LEN_TEST}')

Наблюдений в "data_test.csv": 71231


#### Слияние тренировочной выборки и набора данных с профилем абонента

Предположим, что в features.csv имеются данные для всех id, которые имеются в тренировочной и тестовой выборках.

Проведем слияние методом пересечения (how='inner') по столбцу id.

In [12]:
train_merge = dd.merge(train_df, features_df, on=['id'], how='inner')

In [13]:
train_feat = train_merge.compute()

In [14]:
train_feat.head()

index_vas       id  vas_id  buy_time_vas  target  index_feat  \
0        140  4130548     2.0    1544389200     0.0        8832   
1        842   540997     1.0    1541365200     0.0       11897   
2        843   540997     4.0    1542574800     1.0       11897   
3        926   563378     2.0    1545598800     0.0       12020   
4       1312       37     1.0    1534712400     0.0        9746   

   buy_time_feat          0           1           2           3           4  \
0     1548018000  11.700029   17.790888    4.429214  -10.646798   -0.988246   
1     1545598800 -96.799971  -69.199112 -108.200786 -112.916798 -113.618246   
2     1545598800 -96.799971  -69.199112 -108.200786 -112.916798 -113.618246   
3     1534107600 -96.799971 -242.079112 -110.740786 -294.686798 -116.158246   
4     1545598800 -96.799971 -408.179112 -110.740786 -460.786798 -116.158246   

           5         6          7         8          9        10         11  \
0  -31.75179   7.85382  29.933903 -3.863189 -14.606532 -2.736081  17.012004   
1 -134.02179 -16.08618 -65.076097 -6.783660 -30.006538 -2.736081  -4.007526   
2 -134.02179 -16.08618 -65.076097 -6.783660 -30.006538 -2.736081  -4.007526   
3 -315.79179 -16.08618 -65.076097 -6.783660 -30.006538 -2.736081  -4.007526   
4 -481.89179 -16.08618 -65.076097 -6.783660 -30.006538 -2.736081  -4.007526   

         12         13        14        15        16        17       18  \
0 -2.558912  18.370873   7.04189 -0.000725 -0.016435 -0.107041 -5.41746   
1 -2.558912 -66.189127 -77.51811 -0.000725 -0.016435 -0.107041 -5.41746   
2 -2.558912 -66.189127 -77.51811 -0.000725 -0.016435 -0.107041 -5.41746   
3 -2.558912 -66.189127 -77.51811 -0.000725 -0.016435 -0.107041 -5.41746   
4 -2.558912 -66.189127 -77.51811 -0.000725 -0.016435 -0.107041 -5.41746   

         19         20         21        22        23        24        25  \
0 -3.178521  -7.270815 -10.744164 -0.094251 -0.001733 -0.009327 -2.082209   
1 -3.178521 -11.400815 -10.744164 -0.094251 -0.001733 -0.009327 -2.082209   
2 -3.178521 -11.400815 -10.744164 -0.094251 -0.001733 -0.009327 -2.082209   
3 -3.178521 -13.940815 -10.744164 -0.094251 -0.001733 -0.009327 -2.082209   
4 -3.178521 -13.940815 -10.744164 -0.094251 -0.001733 -0.009327 -2.082209   

         26       27        28        29        30        31        32  \
0  0.200138 -0.00909 -0.351862 -0.214366 -0.211608 -0.001884 -0.000023   
1  0.200138 -0.00909 -0.351862 -0.214366 -0.211608 -0.001884 -0.000023   
2  0.200138 -0.00909 -0.351862 -0.214366 -0.211608 -0.001884 -0.000023   
3  0.200138 -0.00909 -0.351862 -0.214366 -0.211608 -0.001884 -0.000023   
4  0.200138 -0.00909 -0.351862 -0.214366 -0.211608 -0.001884 -0.000023   

        33        34        35        36          37           38  \
0 -0.00003  20.71061 -0.065583 -0.700765  -43.026959  -598.379904   
1 -0.00003  -2.65939 -0.065583 -0.700765 -192.026959 -2922.876926   
2 -0.00003  -2.65939 -0.065583 -0.700765 -192.026959 -2922.876926   
3 -0.00003  -2.65939 -0.065583 -0.700765 -192.026959 -2942.440404   
4 -0.00003  -2.65939 -0.065583 -0.700765 -192.026959 -2942.440404   

            39           40         41          42          43         44  \
0 -1068.177947   469.798136 -89.504287 -119.724355  -47.728685 -23.919190   
1 -1186.295133 -1736.581791 -89.504287 -119.724355 -118.428686 -54.602524   
2 -1186.295133 -1736.581791 -89.504287 -119.724355 -118.428686 -54.602524   
3 -1186.765837 -1755.674564 -89.504287 -119.724355  -99.378684 -34.369190   
4 -1186.765837 -1755.674564 -89.504287 -119.724355 -119.662019 -54.602524   

          45        46        47         48         49         50         51  \
0 -23.809498 -7.896282 -6.822342 -10.707745 -28.565596 -17.130284 -21.088986   
1 -63.826161 -7.896282 -7.997875 -10.717958 -28.571103 -27.130284 -31.088986   
2 -63.826161 -7.896282 -7.997875 -10.717958 -28.571103 -27.130284 -31.088986   
3 -65.009494 -7.896282 -7.997875 -10.717958 -28.571103 -36.130284 -41.088986   
4 -65.059494 

In [15]:
LEN_TRAIN_FEAT = train_feat.shape[0]
print(f'Наблюдений в датасете после слияния": {LEN_TRAIN_FEAT}')

Наблюдений в датасете после слияния": 860052


После слияния тренировочный датасет имеет больше наблюдей (860052), чем исходный (831653). 

* Это связано с тем, что для некоторых клиентов в данных профиля потребления имеются несколько записей, зафиксированных в разное время. 
* Для таких случаев произошло дублирование некоторых наблюдений из data_train.csv с разными записями признков из features.csv. Дублированными строками считаются те, которые имеют одинаковое значение index_vas.

Избавляться от дубликатов будем с учетом актуальности по временни. 
* Среди дубликатов будем оставлять только те наблдения, у которых разница между временными штампами отклика на услугу ('buy_time_vas') и фиксации записи по признакам профиля потребления ('buy_time_feat') минимальна. 
* Введем новый столбец 'time_delta', который будет отражать указанную разницу во времени.

In [16]:
def buy_time_clear(df):
    df['time_delta'] = abs(df['buy_time_vas'] - df['buy_time_feat'])
    
    # Отсортируем наблюдения в порядке возрастания 'time_delta'
    df.sort_values(['time_delta'], inplace=True, ignore_index=True)
    
    # удалим по порядку все дубликаты, оставляя лишь те, которые встречаются в датафрейме первыми (keep='first')
    df.drop_duplicates(['index_vas'], keep='first', inplace=True)
    
    # Отсортируем датасет по возрастанию index_vas.
    df.sort_values(['index_vas'], inplace=True, ignore_index=True)
    
    return df    

In [17]:
buy_time_clear(train_feat)

index_vas       id  vas_id  buy_time_vas  target  index_feat  \
0               0   540968     8.0    1537131600     0.0     3756522   
1               1  1454121     4.0    1531688400     0.0     2735973   
2               2  2458816     1.0    1534107600     0.0     3586550   
3               3  3535012     5.0    1535922000     0.0     1634974   
4               4  1693214     1.0    1535922000     0.0      716515   
...           ...      ...     ...           ...     ...         ...   
831648     831648  3812226     2.0    1546203600     0.0     3598646   
831649     831649  2480469     2.0    1546203600     0.0     1126545   
831650     831650   158236     2.0    1546203600     0.0     3755060   
831651     831651  1825525     2.0    1546203600     0.0      253884   
831652     831652  3022610     2.0    1546203600     0.0     2556316   

        buy_time_feat           0           1           2           3  \
0          1541970000  -31.559971  327.360888  -45.500786  274.753202   
1          1531083600  547.270029  238.430888  533.329214  274.803202   
2          1543179600  -92.139971  -95.469112 -106.080786 -139.596798   
3          1533502800   54.880029   12.970888   54.079214   -9.116798   
4          1543179600   45.160029  295.240888   64.679214  344.283202   
...               ...         ...         ...         ...         ...   
831648     1532898000   29.750029    6.200888   24.279214  -37.936798   
831649     1532293200  -89.179971  -56.499112 -103.120786 -109.106798   
831650     1538946000  -96.799971   62.140888 -110.740786    9.533202   
831651     1533502800  -96.799971  -81.919112 -110.740786 -128.596798   
831652     1540760400  -49.339971 -218.339112   73.139214  145.093202   

                 4          5         6           7         8          9  \
0       -50.918246  253.64821  49.15382  195.883903 -1.835267 -12.831245   
1       527.911754  253.69821 -16.08618  -65.076097 -6.783660 -30.006538   
2      -111.498246 -142.90179 -16.08618  -65.076097 -6.783660 -30.006538   
3        48.661754  -30.22179  60.17382   87.453903 -6.783660 -30.006538   
4        59.261754  323.17821 -16.08618  -65.076097 -6.783660 -30.006538   
...            ...        ...       ...         ...       ...        ...   
831648   18.861754  -59.04179 -16.08618  -65.076097 -6.783660 -30.006538   
831649 -108.538246 -130.21179 -16.08618  -65.076097 -6.783660 -30.006538   
831650 -116.158246  -11.57179 -16.08618  405.243903 -6.783660 -30.006538   
831651 -116.158246  -76.82179 -16.08618  -65.076097 -6.783660 -30.006538   
831652   67.721754  123.98821 -16.08618  -65.076097 -6.783660 -30.006538   

               10         11        12          13         14        15  \
0       -2.736081  56.284081 -2.558912  -66.189127  -77.51811 -0.000725   
1       -2.736081  -4.007526 -2.558912  577.880873  566.55189 -0.000725   
2       -2.736081  -4.007526 -2.558912  -66.189127  -74.97811 -0.000725   
3       73.523919  -4.007526 -2.558912   -9.829127   -2.09811 -0.000725   
4       -2.736081  -4.007526 -2.558912  -66.189127   58.08189 -0.000725   
...           ...        ...       ...         ...        ...       ...   
831648  -2.736081  -4.007526 -2.558912  -66.189127   46.91189 -0.000725   
831649  -2.736081  -4.007526 -2.558912  -66.189127  -77.51811 -0.000725   
831650  -2.736081  -4.007526 -2.558912  -66.189127  -77.51811 -0.000725   
831651  -2.736081  -4.007526 -2.558912  -66.189127  -77.51811 -0.000725   
831652  -2.736081  -4.007526 -2.558912  -18.729127  -30.05811 -0.000725   

              16        17       18        19          20         21  \
0      -0.016435 -0.107041 -5.41746 -3.178521  -13.940815  54.495836   
1      -0.016435 -0.107041 -5.41746 -3.178521  -13.940815 -10.744164   
2      -0.016435 -0.107041 -5.41746 -1.058521  -13.940815 -10.744164   
3      -0.016435 -0.107041 -5.41746 -3.178521   -0.800815 -10.744164   
4      -0.016435 -0.107041 -5.41746  3.181479   19.519185 -10.744164   
...          ...    

Проверим нарушилась ли индексация index_vas путем сравнения со сброшенным на предыдущем этапе индеском датасета.

In [18]:
(train_feat.index == train_feat['index_vas']).all()

True

Индексы совпадают, индексация не нарушена.

Сравним количество наблюдений полученного датасета с data_train.csv

In [19]:
print(f'Наблюдений   в   исходном  "data_train.csv": {LEN_TRAIN}')
print(f'Наблюдений в новом тренирвоочном датафрейме: {train_feat.shape[0]}')

Наблюдений   в   исходном  "data_train.csv": 831653
Наблюдений в новом тренирвоочном датафрейме: 831653


Индексация и количество наблюдений после извлечения и добавления признаков потребления в новом тренировочном датасете НЕ НАРУШЕНЫ.

Сохранение нового тренировочного датасета:
* Удалим лишние стобцы с индексами и признаком buy_time из features.csv. 
* Признак time_delta оставим, так как он несет в себе информацию об актуальности признаков профиля потребления на момент отклика клиента на услугу. 
* Сохраним датасет в data_feat_train.pkl.

In [20]:
train_feat.drop(['index_vas','index_feat', 'buy_time_feat'], axis=1).astype('float32').to_pickle('data_feat_train.pkl')

#### Слияние тестовой выборки и набора данных с профилем абонента

In [21]:
test_merge = dd.merge(test_df, features_df, on=['id'], how='inner')
test_feat = test_merge.compute()

In [22]:
buy_time_clear(test_feat)

index_vas       id  vas_id  buy_time_vas  index_feat  buy_time_feat  \
0              0  3130519     2.0    1548018000     1871362     1536526800   
1              1  2000860     4.0    1548018000     2024591     1532293200   
2              2  1099444     2.0    1546808400     4041332     1541365200   
3              3  1343255     5.0    1547413200     1945062     1536526800   
4              4  1277040     2.0    1546808400     3503668     1532293200   
...          ...      ...     ...           ...         ...            ...   
71226      71226  2502453     5.0    1548018000      820229     1534712400   
71227      71227  1693213     2.0    1548018000     3872196     1541970000   
71228      71228  1891350     2.0    1548018000     3188331     1545598800   
71229      71229  2437172     2.0    1548018000     1510984     1532293200   
71230      71230   988236     2.0    1548018000     1449674     1535922000   

                0           1           2           3           4          5  \
0      -62.899971 -374.279112  -72.600786 -418.406798  -78.018246 -439.51179   
1      -96.799971  100.290888 -110.740786  140.903202 -116.158246  119.79821   
2      -81.969971 -390.729112  -95.910786 -443.336798 -101.328246 -464.44179   
3      259.130029  -52.249112  245.189214 -104.856798  239.771754 -125.96179   
4      331.170029  590.890888  317.229214  538.283202  311.811754  522.25821   
...           ...         ...         ...         ...         ...        ...   
71226  -96.799971 -408.179112 -110.740786 -460.786798 -116.158246 -481.89179   
71227  -89.689971 -355.809112 -103.630786 -408.416798 -109.048246 -429.52179   
71228  -96.799971 -281.059112 -110.740786 -333.666798 -116.158246 -354.77179   
71229  115.060029  -87.339112  101.119214 -127.236798   95.701754 -148.34179   
71230  -96.799971   68.510888 -110.740786   15.903202 -116.158246   83.76821   

              6          7        8          9         10        11        12  \
0      17.81382 -31.176097 -6.78366 -30.006538  31.163919 -4.007526 -2.558912   
1     -16.08618 -65.076097 -6.78366 -30.006538  -2.736081 -4.007526 -2.558912   
2     -16.08618 -65.076097 -6.78366 -30.006538  -2.736081 -4.007526 -2.558912   
3     -16.08618 -65.076097 -6.78366 -30.006538  -2.736081 -4.007526 -2.558912   
4     -16.08618 -65.076097 -6.78366 -30.006538  -2.736081 -4.007526 -2.558912   
...         ...        ...      ...        ...        ...       ...       ...   
71226 -16.08618 -65.076097 -6.78366 -30.006538  -2.736081 -4.007526 -2.558912   
71227 -16.08618 -65.076097 -6.78366 -30.006538  -2.736081 -4.007526 -2.558912   
71228 -16.08618 -65.076097 -6.78366 -30.006538  -2.736081 -4.007526 -2.558912   
71229 -16.08618 -65.076097 -6.78366 -30.006538  -2.736081 -4.007526 -2.558912   
71230 -16.08618 -65.076097 -6.78366 -30.006538  -2.736081 -4.007526 -2.558912   

               13         14        15        16        17       18        19  \
0      -66.189127  -77.51811 -0.000725 -0.016435 -0.107041 -5.41746 -3.178521   
1      -66.189127  -77.51811 -0.000725 -0.016435 -0.107041 -5.41746 -3.178521   
2      -66.189127  -64.80811 -0.000725 -0.016435 -0.107041 -5.41746 -1.058521   
3      289.740873  278.41189 -0.000725 -0.016435 -0.107041 -5.41746 -3.178521   
4      357.540873  350.45189 -0.000725 -0.016435 -0.107041 -5.41746 -3.178521   
...           ...        ...       ...       ...       ...      ...       ...   
71226  -66.189127  -77.51811 -0.000725 -0.016435 -0.107041 -5.41746 -3.178521   
71227  -66.189127  -70.40811 -0.000725 -0.016435 -0.107041 -5.41746 -3.178521   
71228  -66.189127  -77.51811 -0.000725 -0.016435 -0.107041 -5.41746 -3.178521   
71229  145.670873  134.34189 -0.000725 -0.016435 -0.107041 -5.41746 -3.178521   
71230  -66.189127  -77.51811 -0.000725 -0.016435 -0.107041 -5.41746 -3.178521   

              20         21        22        23        24        25        26  \
0      -9.700815 -10.744164 -0.094251 -0.001733 -0.009327 -2.082209  0.200138   
1

In [23]:
(test_feat.index == test_feat['index_vas']).all()

True

In [24]:
print(f'Наблюдений в исходном  "data_test.csv": {LEN_TEST}')
print(f'Наблюдений в новом тестовом датафрейме: {test_feat.shape[0]}')

Наблюдений в исходном  "data_test.csv": 71231
Наблюдений в новом тестовом датафрейме: 71231


Индексация и количество наблюдений после извлечения и добавления признаков потребления в новом тестовом датасете НЕ НАРУШЕНЫ.

Сохраним датафрейм в data_feat_test.pkl.

In [25]:
test_feat.drop(['index_vas','index_feat'], axis=1).astype('float32').to_pickle('data_feat_test.pkl')